---
title: "Inferring Presence or Absence of Sex Chromosomes"
author: "Seema Plaisier and Mariah Lee"
date: " Last Updated `r format(Sys.time(), '%m/%d/%y')`"
output: 
  pdf_document: default
  html_document:
    df_print: paged
editor_options:
  markdown:
    wrap: 60
---

## Print Options

```{r setup, include=FALSE}
# Set the root directory for knitting the document
knitr::opts_knit$set(root.dir = "D:/CCLE_Project/Infer_Y_output/")
# Set chunk options to improve readability of code output
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
```

## Overview

The purpose of this file is to infer the presence or absence of the X and Y chromosomes in the cell lines in the CCLE. XIST will be used to infer the presence of one or two X chromosomes (or one active X chromosome and one inactive chromosome). Multiple X-degenerate genes will be used to infer the presence of a Y chromosome as well as possible deletion or down regulation.

## Install and Load Needed Packages

```{r LoadPackages, message=FALSE, warning=FALSE}
# Function to install and load packages
load_or_install <- function(pkg) {
  if (!require(pkg, character.only = TRUE)) {
    install.packages(pkg)
    library(pkg, character.only = TRUE)
  }
}

# List of required packages
packages <- c("dplyr", "tidyr", "ggplot2")

# Install and load each package
sapply(packages, load_or_install)
```

## Set Up Data and Working Directories

```{r Directories, message=FALSE}
# Directory where the data is stored; include / at the end
data_directory <- "D:/CCLE_Project/data/"
```

## Read Data

The gene expression matrix file contains a table with all the genes as rows and all the cell lines in the CCLE as columns, plus a few extra columns giving important information about the genes like name and IDs in important databases. It is is fairly large and will likely take a few minutes to load.

The sample annotation file contains information about each of the cell lines in the CCLE. This includes tumor tissue type the cell line was derived from, patient information like age and "gender" (reported sex), as well as some information about how the cell line was grown prior to sample collection for RNA sequencing.

```{r ReadData, message=FALSE}
# Check if data files exist before loading
if (file.exists(paste0(data_directory,"CCLE_RNAseq_genes_counts_20180929.csv"))) {
  CCLE_data <- read.csv(paste0(data_directory,"CCLE_RNAseq_genes_counts_20180929.csv"), header = TRUE)
} else {
  stop("CCLE RNAseq gene expression file not found.")
}

if (file.exists(paste0(data_directory,"Cell_lines_annotations_20181226.txt"))) {
  annotation_data <- read.csv(paste0(data_directory,"Cell_lines_annotations_20181226.txt"), header = TRUE, sep = "\t")
} else {
  stop("Cell line annotation file not found.")
}

# load the CCLE gene expression data
CCLE_data = read.csv(paste0(
  data_directory,"CCLE_RNAseq_genes_counts_20180929.csv"), header = TRUE)

# load the annotation data
annotation_data = read.csv(paste0(
  data_directory,"Cell_lines_annotations_20181226.txt"), header = TRUE, sep = "\t")
```

## Look-Up Tables

To associate our cell lines with annotation information, we can created named lists to serve as look up table. In these tables, we will pull all the cells with the annotation we are interested in, and name them with the cell line so we can quickly look it up below.

```{r LookUp, message=FALSE}
# Create a look-up table for the reported sex based on CCLE_ID

# Extract reported sex from annotation data
get_reported_sex <- annotation_data$Gender 

# Name entries by CCLE_ID
names(get_reported_sex) <- annotation_data$CCLE_ID

# Remove NA values indicating no reported sex
get_reported_sex <- na.omit(get_reported_sex)

# Display unique levels of reported sex
levels(factor(get_reported_sex))

# Filter for "female" and "male" only
get_reported_sex <- get_reported_sex[get_reported_sex %in% c("female", "male")]
```

## Gene of Interest

In this chunk, we will set the name of the gene we are interested in analyzing. You will need to make sure that the gene name you choose is in the 'Description' column of the annotation_data variable which contains the gene expression matrix.

Here we use the 'subset' function to pull out the row with the selected gene name in the Description column. We print what we found to make sure that we actually got what we wanted. If you do not see any information in the 'chosen_gene_data' variable, either the gene name is spelled wrong or that gene is not included in the data set under that name. It might be listed using an alternative gene name.

```{r ChooseGene}
# Set the gene of interest
chosen_gene <- "SRY"

# Subset the gene expression data by the chosen gene
chosen_gene_data <- subset(CCLE_data, Description == chosen_gene)

# Check if the chosen gene is found in the dataset
if (nrow(chosen_gene_data) == 0) {
  stop("Chosen gene not found in the dataset. Please check the gene name.")
}

# Display the first few entries to verify data
print(head(chosen_gene_data))

# Remove unnecessary columns (Name and Description)
chosen_gene_subset <- subset(chosen_gene_data, select = -c(1,2))
```

## Gene expression in cell lines with different reported sex

Here we want to look at the expression of the chosen gene in cell lines annotated to have 'female' or 'male' as their reported sex. 

```{r ViewDataBySex}
# Filter expression data for cell lines with reported sex
chosen_gene_reported_sex <- chosen_gene_subset[colnames(
  chosen_gene_subset) %in% names(get_reported_sex)]

# Transpose, log-transform, and convert to data frame
chosen_gene_reported_sex <- as.data.frame(t(log(chosen_gene_reported_sex)))

# Set column name to "expression"
colnames(chosen_gene_reported_sex) <- "expression"

# Replace -Inf values with NA
chosen_gene_reported_sex$expression[chosen_gene_reported_sex$expression == "-Inf"] <- NA

# Add reported sex column using vectorization instead of a loop
chosen_gene_reported_sex$reported_sex <- get_reported_sex[rownames(chosen_gene_reported_sex)]
```

## Violin Plots

We would like to see the range of values we see for the expression of the gene-- do all cell lines from patients reported as female have high expression and as male have low? What threshold do we use to even say what is high or low expression?

```{r ViolinPlots}
# Calculate metrics used for thresholds
female_mean = mean(chosen_gene_reported_sex$expression
                    [chosen_gene_reported_sex$reported_sex=='female'],
                    na.rm = TRUE)
female_sd = sd(chosen_gene_reported_sex$expression
                [chosen_gene_reported_sex$reported_sex=='female'], 
                na.rm = TRUE)

# add line to think about thresholds that can be used to predict reported sex
high_threshold = female_mean + (4.5*female_sd)
low_threshold = female_mean +(3*female_sd)

# use ggplot to visualize the data as a violin-jitter plot
violin_with_thresholds = ggplot(chosen_gene_reported_sex, 
      aes(x=reported_sex, y=chosen_gene_reported_sex$expression, fill=reported_sex)) + 
      geom_violin(trim = FALSE) + 
      scale_fill_manual(values=c("grey", "orange"))+ 
      geom_jitter(size = 0.75) + 
      ylab(paste0(chosen_gene," Expression (log counts)")) + 
      geom_hline(yintercept = high_threshold, linetype="dashed", color = "maroon") + 
      geom_hline(yintercept = low_threshold, linetype="dashed", color = "blue") + 
      theme_light()

# Save violin plot with thresholds as SVG
ggsave(paste0("violin_plot_image_",chosen_gene,".svg"), 
       violin_with_thresholds, width = 13, height = 9, units = c("cm"), device = "svg")
```

## Pie Charts

Here we will make pie charts that show the amount of cell lines with expression above and below the high and low thresholds. There will be one pie chart for cell lines reported from females and one pie chart from cell lines reported from males.

```{r PieCharts}
# Function to create pie chart and save as SVG
create_pie_chart <- function(data, reported_sex) {
  num_over_high <- nrow(subset(data, reported_sex == reported_sex & 
                                 expression >= high_threshold))
  num_between <- nrow(subset(data, reported_sex == reported_sex & 
                               expression < high_threshold & 
                               expression > low_threshold))
  num_under_low <- nrow(subset(data, reported_sex == reported_sex & 
                                 expression <= low_threshold))
  
  nums <- c(num_over_high, num_between, num_under_low)
  labels <- c(paste0(">= high threshold of ", high_threshold), "between",
              paste0("<= low threshold of ", low_threshold))
  
  # Save pie chart as SVG
  svg(paste0("pie_chart_", chosen_gene, "_", reported_sex, ".svg"), width = 7, height = 7)
  pie(nums, labels = labels, main = paste0("Thresholds for cell lines reported ",
                                           reported_sex, " [n = ", sum(nums), "]"), 
      col = c("orange", "grey", "blue"))
  dev.off()
}

# Create pie charts for female and male reported sex
create_pie_chart(chosen_gene_reported_sex, "female")
create_pie_chart(chosen_gene_reported_sex, "male")
```

## Expression Table

Here we will use the thresholds we came up with to categorize the expression levels of each cell line.

```{r ExpressionTable}

# use chosen thresholds to predict sex based on expression

# start with all the cell lines that have expression of the gene you are using to make a prediction
# make a data frame containing the expression
predict_sex = data.frame(t(chosen_gene_subset))
colnames(predict_sex) = "expression"

# add column for expression threshold
predict_sex$predicted_sex = vector(length = nrow(predict_sex)) # empty placeholders 
for (cell_line in rownames(predict_sex)) {
  if (log(predict_sex[cell_line,"expression"]) >= high_threshold) {
    predict_sex[cell_line,"predicted_sex"] = "high_expression"
  } 
  else if (log(predict_sex[cell_line,"expression"]) <= low_threshold) {
    predict_sex[cell_line,"predicted_sex"] = "low_expression"
  } 
  else if (log(predict_sex[cell_line,"expression"]) > low_threshold & 
             log(predict_sex[cell_line,"expression"]) < high_threshold) {
    predict_sex[cell_line,"predicted_sex"] = "intermediate_expression"
  }
}

# add in the reported sex if present
predict_sex$reported_sex = vector(length = nrow(predict_sex)) # empty placeholders 
for (cell_line in rownames(predict_sex)) {
  predict_sex[cell_line,"reported_sex"] = get_reported_sex[cell_line]
}

# add log values and indicate NA values
log_expression = log(predict_sex$expression)
log_expression[log_expression == "-Inf"] = NA
predict_sex$log_expression = log_expression

# reorder columns
predict_sex <- predict_sex[ , c(3,1,4,2)]

# add the name of the gene you used to predict for good housekeeping
expression_index = which( colnames(predict_sex)== "expression" )
colnames(predict_sex)[expression_index] = paste0(chosen_gene,"_expression")
predicted_sex = which( colnames(predict_sex)== "predicted_sex" )
colnames(predict_sex)[predicted_sex]= paste0(chosen_gene,"_expression_level")
log_expression = which( colnames(predict_sex)== "log_expression" )
colnames(predict_sex)[log_expression] = paste0(chosen_gene,"_log_expression")

# write to output file so we can look at it later
predicted_sex_output_file = paste0("predicted_sex_",chosen_gene,".csv")
write.csv(predict_sex, file = predicted_sex_output_file)
```

## List all the packages used for future reference

```{r SessionInfo}
sessionInfo()
```
