---
title: "Inferring Presence or Absence of Sex Chromosomes"
author: "Seema Plaisier and Mariah Lee"
date: "`r format(Sys.time(), '%m/%d/%y')`"
output: 
  pdf_document: default
  html_document:
    df_print: paged
editor_options:
  markdown:
    wrap: 60
---

## Overview

The purpose of this file is to infer the presence or absence
of the Y chromosome in cell lines of the CCLE. Multiple
X-degenerate genes will be used to infer the presence of a Y
chromosome as well as possible deletion or down regulation.

## Install and Load Needed Packages

```{r LoadPackages, message=FALSE, warning=FALSE}

# check if the package has been installed; if not, install it
if(!require(dplyr)){
    install.packages("dplyr")
}
if(!require(tidyr)){
    install.packages("tidyr")
}
if(!require(ggplot2)){
    install.packages("ggplot2")
}
if(!require(knitr)){
    install.packages("knitr")
}
# load packages
library(dplyr)          # data wrangling
library(tidyr)          # data tidying
library(ggplot2)        # making plots
library(knitr)          # neat printing

```

## Print Options

```{r setup, include=FALSE}
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
```

## Set Up Data and Working Directories

```{r Directories, message=FALSE}
 
# set working directory to where you want your outputs to go
    # make sure to include / at the end of directories
working_directory <- "D:/GDrive_Dup/" 
setwd(working_directory)

# this is where the data is stored
data_directory <- "D:/GDrive_Dup/"

```

## Import Data

The gene expression data file contains a table with all the
genes as rows and all the cell lines in the CCLE as columns,
plus a few extra columns giving important information about
the genes like name and IDs in important databases. It is is
fairly large and will likely take a few minutes to load.

The sample annotation file contains information about each
of the cell lines in the CCLE. This includes tumor tissue
type the cell line was derived from, patient information
like age and "gender" (reported sex), as well as some
information about how the cell line was grown prior to
sample collection for RNA sequencing.

```{r ReadData, message=FALSE}

# load the CCLE gene expression data
CCLE_data = read.csv(paste0(data_directory,"CCLE_RNAseq_genes_counts_20180929.csv"), header = TRUE)

# load the annotation data
annotation_data = read.csv(paste0(data_directory,"Cell_lines_annotations_20181226.txt"), header = TRUE, sep = "\t")

```

## Look-Up Tables

To associate our cell lines with annotation information, we
can created named lists to serve as look up table. In these
tables, we will pull all the cells with the annotation we
are interested in, and name them with the cell line so we
can quickly look it up below.

```{r LookUp, message=FALSE}
#   this code will make a quick lookup table to get the reported age of the patient when the cell line was started

# pull out the Gender column of the annotation data frame
get_reported_sex = annotation_data$Gender 

# use the CCLE_ID to label each entry of the reported sex
# this creates a lookup table
names(get_reported_sex) = annotation_data$CCLE_ID

# remove NA values which indicate that no sex was reported
get_reported_sex = na.omit(get_reported_sex)

# see what values you got from the reported sex in the Gender column
levels(factor(get_reported_sex))

# filter for only the entries that are reported as "female" or "male"
get_reported_sex = get_reported_sex[get_reported_sex == "female" | get_reported_sex == "male"] # filter
```

## Gene of Interest

In this chunk, we will set the name of the gene we are
interested in analyzing. You will need to make sure that the
gene name you choose is in the 'Description' column of the
annotation_data variable which contains the gene expression
matrix.

Here we use the 'subset' function to pull out the row with
the selected gene name in the Description column. We print
what we found to make sure that we actually got what we
wanted. If you do not see any information in the
'chosen_gene_data' variable, either the gene name is spelled
wrong or that gene is not included in the data set under
that name. It might be listed using an alternative gene
name.

```{r ChooseGene}

# set the gene that you are interested in
chosen_gene = "RPS4Y1"

# pull out the data for this gene from the full data matrix
chosen_gene_data = subset(CCLE_data, Description == chosen_gene)

# print the first few entries just to make sure you have data
# if not, you might have indicated a gene name that is not present in the data set
# which could be due to incorrect spelling or capitalization
print (chosen_gene_data[1:10])

# remove the first two columns (Name and Description),
# since we are only want to plot expression data values
chosen_gene_subset = subset(chosen_gene_data, select = -c(1,2))

```

## Gene expression in cell lines with different reported sex

Here we want to look at the expression of the chosen gene in
cell lines annotated to have 'female' or 'male' as their
reported sex.

```{r ViewDataBySex}

# pull out expression data for cell lines that had a reported sex of female or male
chosen_gene_reported_sex = chosen_gene_subset[colnames(chosen_gene_subset) %in% names(get_reported_sex)]

# transpose and log transform the data so it's easier to work with and put it into a data frame
chosen_gene_reported_sex = as.data.frame(t(log(chosen_gene_reported_sex)))

# set the name of the column to be "expression" so we know what to call it when we plot
colnames(chosen_gene_reported_sex) = "expression"

# change -Inf values to NA values
chosen_gene_reported_sex$expression[chosen_gene_reported_sex$expression == "-Inf"] = NA

#  add a reported sex column
chosen_gene_reported_sex$reported_sex = vector(length = nrow(chosen_gene_reported_sex)) # empty placeholders 
for (cell_line in rownames(chosen_gene_reported_sex)) {
  chosen_gene_reported_sex[cell_line,"reported_sex"] = get_reported_sex[cell_line]
}

# descriptive statistics in males and females
chosen_gene_male_data = chosen_gene_reported_sex[
  chosen_gene_reported_sex$reported_sex == "male", ]
chosen_gene_female_data = chosen_gene_reported_sex[
  chosen_gene_reported_sex$reported_sex == "female", ]

chosen_gene_male_mean = mean(chosen_gene_male_data$expression, na.rm = TRUE)
chosen_gene_male_SD = sd(chosen_gene_male_data$expression, na.rm = TRUE)

chosen_gene_female_mean = mean(chosen_gene_female_data$expression, na.rm = TRUE)
chosen_gene_female_SD = sd(chosen_gene_female_data$expression, na.rm = TRUE)

```

# Violin Plots

We would like to see the range of values we see for the
expression of the gene-- do all cell lines from patients
reported as female have high expression and as male have
low? What threshold do we use to even say what is high or
low expression?

```{r ViolinPlots}

# use ggplot to visualize the data as a violin-jitter plot
violin_plot <- ggplot(chosen_gene_reported_sex, 
      aes(x=reported_sex, 
          y=chosen_gene_reported_sex$expression, 
          fill=reported_sex)) + 
      geom_violin(trim = FALSE) + 
      scale_fill_manual(values=c("grey", "orange"))+ 
      geom_jitter(size = 0.75) + 
      ylab(paste0(chosen_gene," Expression (log counts)")) + 
      theme_light()

# add line to think about thresholds that can be used to predict reported sex
high_threshold = chosen_gene_female_mean + (4.5*chosen_gene_female_SD)

low_threshold = chosen_gene_female_mean + (3*chosen_gene_female_SD)

violin_with_thresholds <- ggplot(chosen_gene_reported_sex, 
      aes(x=reported_sex, y=chosen_gene_reported_sex$expression, fill=reported_sex)) + 
      geom_violin(trim = FALSE) + 
      scale_fill_manual(values=c("grey", "orange"))+ 
      geom_jitter(size = 0.75) + 
      ylab(paste0(chosen_gene," Expression (log counts)")) + 
      geom_hline(yintercept = high_threshold, linetype="dashed", color = "maroon") + 
      geom_hline(yintercept = low_threshold, linetype="dashed", color = "blue") + 
      theme_light()

# save violin plot with thresholds as image
ggsave(paste0("violin_plot_ ",chosen_gene,".png"), 
       violin_with_thresholds, width = 13, height = 9, units = c("cm"))

```

## Pie Charts

Here we will make pie charts that show the amount of cell
lines with expression above and below the high and low
thresholds. There will be one pie chart for cell lines
reported from females and one pie chart from cell lines
reported from males.

```{r PieCharts}
# count how many cell lines reported female are in each region defined by the thresholds chosen
num_females_over_high_threshold = nrow(subset(chosen_gene_reported_sex, 
                                              reported_sex == "female" & 
                                                expression >= high_threshold))
num_females_between_thresholds = nrow(subset(chosen_gene_reported_sex, 
                                             reported_sex == "female" & 
                                               expression < high_threshold & 
                                               expression > low_threshold))
num_females_under_low_threshold = nrow(subset(chosen_gene_reported_sex, 
                                              reported_sex == "female" & 
                                                expression <= low_threshold))

# pie chart to show proportions using the chosen threshold
# can use this to try a few different thresholds
female_threshold_nums = c(num_females_over_high_threshold, 
                          num_females_between_thresholds, 
                          num_females_under_low_threshold)
female_threshold_labels = c(paste0(">= high threshold of ",
                                   round(high_threshold, 2)),"between",
                            paste0("<= low threshold of ",
                                   round(low_threshold, 2)))

# save the pie chart as an image
png(paste0("piechart_female_",chosen_gene,".png"),
    width = 15, height = 8, units = "cm", res = 1080)

piechart_female <- pie(female_threshold_nums, labels = female_threshold_labels, 
    main = paste0 ("Thresholds for cell lines reported female [n =",
                   sum(female_threshold_nums),"]"), 
    col = c("orange","grey","blue"))

dev.off()

# now do the same for cell lines reported as male
num_males_over_high_threshold = nrow(subset(chosen_gene_reported_sex, 
                                            reported_sex == "male" & 
                                              expression >= high_threshold))
num_males_between_thresholds = nrow(subset(chosen_gene_reported_sex, 
                                           reported_sex == "male" & 
                                             expression < high_threshold & 
                                             expression > low_threshold))
num_males_under_low_threshold = nrow(subset(chosen_gene_reported_sex, 
                                            reported_sex == "male" & 
                                              expression <= low_threshold))

male_threshold_nums = c(num_males_over_high_threshold, 
                        num_males_between_thresholds, 
                        num_males_under_low_threshold)

male_threshold_labels = c(paste0(">= high threshold of ",
                                 round(high_threshold, 2)),"between",
                          paste0("<= low threshold of ",
                                 round(low_threshold, 2)))

# save pie chart as an image
png(paste0("piechart_male_",chosen_gene,".png"),
    width = 13, height = 8, units = "cm", res = 1080)

piechart_male <- pie(male_threshold_nums, labels = male_threshold_labels, 
    main = paste0 ("Thresholds for cell lines reported male [n =",
                   sum(male_threshold_nums),"]"), 
    col = c("orange","grey","blue"))

dev.off()

```

## Expression Table

Here we will use the thresholds we came up with to
categorize the expression levels of each cell line.

```{r ExpressionTable}

# use chosen thresholds to predict sex based on expression

# start with all the cell lines that have expression of the gene you are using to make a prediction
# make a data frame containing the expression
predict_sex = data.frame(t(chosen_gene_subset))
colnames(predict_sex) = "expression"

# add column for predicted sex
#  NOTE: how this prediction is done should not be the same for all genes

predict_sex$predicted_sex = vector(length = nrow(predict_sex)) # empty placeholders 
for (cell_line in rownames(predict_sex)) {
  if (log(predict_sex[cell_line,"expression"]) >= high_threshold) {
    predict_sex[cell_line,"predicted_sex"] = "high_expression"
  } 
  else if (log(predict_sex[cell_line,"expression"]) <= low_threshold) {
    predict_sex[cell_line,"predicted_sex"] = "low_expression"
  } 
  else if (log(predict_sex[cell_line,"expression"]) > low_threshold & 
             log(predict_sex[cell_line,"expression"]) < high_threshold) {
    predict_sex[cell_line,"predicted_sex"] = "intermediate_expression"
  }
}

# add in the reported sex if present
predict_sex$reported_sex = vector(length = nrow(predict_sex)) # empty placeholders 
for (cell_line in rownames(predict_sex)) {
  predict_sex[cell_line,"reported_sex"] = get_reported_sex[cell_line]
}
# add log values and indicate NA values
predict_sex["log_expression"] <- log(predict_sex$expression)
predict_sex$log_expression[predict_sex$log_expression == "-Inf"] = NA

# Reorder columns
predict_sex <- predict_sex[ , c(3,1,4,2)]

# Add the name of the gene you used to predict for good housekeeping
expression_index = which( colnames(predict_sex)== "expression" )
colnames(predict_sex)[expression_index] = paste0(chosen_gene,"_expression")
predicted_sex = which( colnames(predict_sex)== "predicted_sex" )
colnames(predict_sex)[predicted_sex]= paste0(chosen_gene,"_expression_category")
log_expression = which( colnames(predict_sex)== "log_expression" )
colnames(predict_sex)[log_expression] = paste0(chosen_gene,"_log_expression")

# write to output file so we can look at it later
predicted_sex_output_file = paste0("predicted_sex_",chosen_gene,"_FINAL.csv")
write.csv(predict_sex, file = predicted_sex_output_file)

```

## List all the packages used for future reference

```{r SessionInfo}
sessionInfo()
```
